# RILanalysis
R scripts for finding and plotting recombination breakpoints in a RIL population

###### Citations
These scripts were used in ![this](https://link.springer.com/article/10.1007/s00122-016-2844-6) 2017 Theoretical and Applied Genetics paper. 
Please cite: Boyles, R.E., Pfeiffer, B.K., Cooper, E.A. et al. Theor Appl Genet (2017) 130: 697. https://doi.org/10.1007/s00122-016-2844-6

Also, these scripts on based on the methods described by Huang et. al (2009).  
Please cite: Huang X, Feng Q, Qian Q, Zhao Q, Wang L, Wang A, Guan J, Fan D, Weng Q, Huang T, Dong G, Sang T, Han B (2009) High-throughput genotyping by whole-genome resequencing. Genome Res 19:1068â€“1076. doi:10.1101/gr.089516.108

### SNP Binning
The first step is to assign the SNPs to "bins"/recombination blocks by finding all transition/breakpoints where the parental haplotype swithches.
Note that the SNPs must be in the format of "A/B/H", where, which is easily generated by TASSEL.

To run the R code in `SNP_Binning_2.R` on your own data, you will need to adjust the following lines:
+  Line 1: Set the path to your working directory
+  Line 71: Enter the name of your input SNP file here
+  Line 87: This loop iterates over the 10 chromosomes in sorghum.  If your species has a different number of chromosomes, you will need to adjust this number.
+  Line 122: Same as line 87; define the chromosome number appropriately.
+  Line 240: Enter the name of the output file you would like to create.

### Plotting
Once you have the breakpoints from the SNP_Binning_2.R script, you can plot the haplotype blocks for all chromosomes.
To use the script `RIL_recomb_plots.R` you only need to change 1 line:
+  Line 7: Enter the correct name for your input file.
